File: ios/BarterApp/BarterApp/Services/ApiService.swift
import Foundation

enum ApiError: Error, LocalizedError {
    case invalidURL
    case noData
    case decodingError(Error)
    case serverError(String)
    case networkError(Error)
    case unauthorized
    
    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "无效的URL"
        case .noData:
            return "没有数据"
        case .decodingError(let error):
            return "数据解析错误: \(error.localizedDescription)"
        case .serverError(let message):
            return message
        case .networkError(let error):
            return "网络错误: \(error.localizedDescription)"
        case .unauthorized:
            return "登录已过期，请重新登录"
        }
    }
}

class ApiService {
    static let shared = ApiService()
    
    private let baseURL: String
    private let session: URLSession
    
    private init() {
        self.baseURL = ApiConfig.shared.baseURL
        self.session = URLSession.shared
    }
    
    // MARK: - Generic Request
    private func request<T: Codable>(
        endpoint: String,
        method: String = "GET",
        body: Encodable? = nil,
        requiresAuth: Bool = true
    ) async throws -> T {
        guard let url = URL(string: "\(baseURL)\(endpoint)") else {
            throw ApiError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        if requiresAuth, let token = TokenManager.shared.token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        
        if let body = body {
            request.httpBody = try JSONEncoder().encode(body)
        }
        
        do {
            let (data, response) = try await session.data(for: request)
            
            if let httpResponse = response as? HTTPURLResponse {
                if httpResponse.statusCode == 401 {
                    throw ApiError.unauthorized
                }
            }
            
            let decoder = JSONDecoder()
            return try decoder.decode(T.self, from: data)
        } catch let error as ApiError {
            throw error
        } catch let error as DecodingError {
            throw ApiError.decodingError(error)
        } catch {
            throw ApiError.networkError(error)
        }
    }
    
    // MARK: - Auth
    func login(username: String, password: String) async throws -> ApiResponse<AuthResponse> {
        let body = LoginRequest(username: username, password: password)
        return try await request(endpoint: "/api/auth/login", method: "POST", body: body, requiresAuth: false)
    }
    
    func register(username: String, email: String, password: String, nickname: String?) async throws -> ApiResponse<AuthResponse> {
        let body = RegisterRequest(username: username, email: email, password: password, nickname: nickname)
        return try await request(endpoint: "/api/auth/register", method: "POST", body: body, requiresAuth: false)
    }
    
    // MARK: - User
    func getProfile() async throws -> ApiResponse<User> {
        return try await request(endpoint: "/api/users/me")
    }
    
    func getUserProfile(userId: Int64) async throws -> ApiResponse<PublicProfile> {
        return try await request(endpoint: "/api/users/\(userId)")
    }
    
    func getAdminUser() async throws -> ApiResponse<PublicProfile> {
        return try await request(endpoint: "/api/users/admin")
    }
    
    func getUserRatings(userId: Int64, page: Int = 0) async throws -> ApiResponse<PageResponse<UserRatingResponse>> {
        return try await request(endpoint: "/api/users/\(userId)/ratings?page=\(page)")
    }
    
    func rateUser(userId: Int64, rating: Int, comment: String?) async throws -> ApiResponse<UserRatingResponse> {
        struct RateRequest: Codable {
            let rating: Int
            let comment: String?
        }
        let body = RateRequest(rating: rating, comment: comment)
        return try await request(endpoint: "/api/users/\(userId)/rate", method: "POST", body: body)
    }
    
    // MARK: - Items
    func getItems(page: Int = 0, size: Int = 20, category: String? = nil, keyword: String? = nil) async throws -> ApiResponse<PageResponse<Item>> {
        var endpoint = "/api/items?page=\(page)&size=\(size)"
        if let category = category, !category.isEmpty {
            endpoint += "&category=\(category)"
        }
        if let keyword = keyword, !keyword.isEmpty {
            endpoint += "&keyword=\(keyword.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? keyword)"
        }
        return try await request(endpoint: endpoint)
    }
    
    func getItemDetail(itemId: Int64) async throws -> ApiResponse<Item> {
        return try await request(endpoint: "/api/items/\(itemId)")
    }
    
    func getMyItems(page: Int = 0) async throws -> ApiResponse<PageResponse<Item>> {
        return try await request(endpoint: "/api/items/my?page=\(page)")
    }
    
    func createItem(title: String, description: String?, category: String?, condition: String, images: [String]?) async throws -> ApiResponse<Item> {
        let body = CreateItemRequest(title: title, description: description, category: category, condition: condition, images: images)
        return try await request(endpoint: "/api/items", method: "POST", body: body)
    }
    
    func deleteItem(itemId: Int64) async throws -> ApiResponse<String> {
        return try await request(endpoint: "/api/items/\(itemId)", method: "DELETE")
    }
    
    // MARK: - Conversations & Messages
    func getConversations(page: Int = 0) async throws -> ApiResponse<PageResponse<Conversation>> {
        return try await request(endpoint: "/api/conversations?page=\(page)")
    }
    
    func getConversationDetail(conversationId: Int64) async throws -> ApiResponse<ConversationDetail> {
        return try await request(endpoint: "/api/conversations/\(conversationId)")
    }
    
    func sendMessage(receiverId: Int64, content: String) async throws -> ApiResponse<SendMessageResponse> {
        let body = SendMessageRequest(receiverId: receiverId, content: content)
        return try await request(endpoint: "/api/messages", method: "POST", body: body)
    }
    
    // MARK: - Trade Requests
    func getTradeRequests(page: Int = 0) async throws -> ApiResponse<PageResponse<TradeRequest>> {
        return try await request(endpoint: "/api/trade-requests?page=\(page)")
    }
    
    func getTradeRequestDetail(tradeId: Int64) async throws -> ApiResponse<TradeRequest> {
        return try await request(endpoint: "/api/trade-requests/\(tradeId)")
    }
    
    func createTradeRequest(targetItemId: Int64, offeredItemId: Int64, message: String?) async throws -> ApiResponse<TradeRequest> {
        struct CreateTradeRequest: Codable {
            let targetItemId: Int64
            let offeredItemId: Int64
            let message: String?
        }
        let body = CreateTradeRequest(targetItemId: targetItemId, offeredItemId: offeredItemId, message: message)
        return try await request(endpoint: "/api/trade-requests", method: "POST", body: body)
    }
    
    func respondTradeRequest(tradeId: Int64, accept: Bool) async throws -> ApiResponse<TradeRequest> {
        let action = accept ? "accept" : "reject"
        return try await request(endpoint: "/api/trade-requests/\(tradeId)/\(action)", method: "POST")
    }
}

File: android/app/src/main/java/com/barter/app/data/remote/ApiService.kt
package com.barter.app.data.remote

import com.barter.app.data.model.*
import okhttp3.MultipartBody
import okhttp3.RequestBody
import retrofit2.Response
import retrofit2.http.*

interface ApiService {

    // ========== 认证 ==========
    @POST("api/auth/login")
    suspend fun login(@Body request: LoginRequest): Response<ApiResponse<AuthResponse>>

    @POST("api/auth/register")
    suspend fun register(@Body request: RegisterRequest): Response<ApiResponse<AuthResponse>>

    // ========== 用户 ==========
    @GET("api/users/me")
    suspend fun getMyProfile(): Response<ApiResponse<User>>

    @GET("api/users/{id}")
    suspend fun getProfile(@Path("id") userId: Long): Response<ApiResponse<User>>

    @PUT("api/users/me")
    suspend fun updateProfile(@Body request: UpdateProfileRequest): Response<ApiResponse<User>>

    @Multipart
    @POST("api/users/me/avatar")
    suspend fun updateAvatar(@Part avatar: MultipartBody.Part): Response<ApiResponse<User>>

    @POST("api/users/me/password")
    suspend fun changePassword(@Body request: ChangePasswordRequest): Response<ApiResponse<Unit>>

    @GET("api/users/me/settings")
    suspend fun getSettings(): Response<ApiResponse<UserSettings>>

    @PUT("api/users/me/settings")
    suspend fun updateSettings(@Body request: UpdateSettingsRequest): Response<ApiResponse<UserSettings>>

    @GET("api/users/me/login-records")
    suspend fun getLoginRecords(): Response<ApiResponse<List<LoginRecord>>>
    
    @GET("api/users/me/ratings")
    suspend fun getMyRatings(): Response<ApiResponse<List<Rating>>>

    @GET("api/users/{id}")
    suspend fun getUserProfile(@Path("id") userId: Long): Response<ApiResponse<PublicProfile>>

    @POST("api/users/{id}/rate")
    suspend fun rateUser(
        @Path("id") userId: Long,
        @Body request: RateUserRequest
    ): Response<ApiResponse<UserRatingResponse>>

    @GET("api/users/{id}/ratings")
    suspend fun getUserRatings(
        @Path("id") userId: Long,
        @Query("page") page: Int = 0,
        @Query("size") size: Int = 20
    ): Response<ApiResponse<PageResponse<UserRatingResponse>>>

    @GET("api/users/admin")
    suspend fun getAdminUser(): Response<ApiResponse<PublicProfile>>

    // ========== 物品 ==========
    @GET("api/items/list")
    suspend fun getItems(
        @Query("page") page: Int = 0,
        @Query("size") size: Int = 20
    ): Response<ApiResponse<PageResponse<ItemListItem>>>

    @GET("api/items/search")
    suspend fun searchItems(
        @Query("keyword") keyword: String,
        @Query("page") page: Int = 0,
        @Query("size") size: Int = 20
    ): Response<ApiResponse<PageResponse<ItemListItem>>>

    @GET("api/items/{id}")
    suspend fun getItem(@Path("id") itemId: Long): Response<ApiResponse<Item>>

    @GET("api/items/my")
    suspend fun getMyItems(
        @Query("page") page: Int = 0,
        @Query("size") size: Int = 20
    ): Response<ApiResponse<PageResponse<ItemListItem>>>

    @GET("api/items/my/available")
    suspend fun getMyAvailableItems(): Response<ApiResponse<List<ItemListItem>>>

    @Multipart
    @POST("api/items")
    suspend fun createItem(
        @Part("item") item: RequestBody,
        @Part images: List<MultipartBody.Part>?
    ): Response<ApiResponse<Item>>

    @PUT("api/items/{id}")
    suspend fun updateItem(
        @Path("id") itemId: Long,
        @Body request: UpdateItemRequest
    ): Response<ApiResponse<Item>>

    @DELETE("api/items/{id}")
    suspend fun deleteItem(@Path("id") itemId: Long): Response<ApiResponse<Unit>>
    
    @POST("api/items/{id}/wish")
    suspend fun toggleWish(@Path("id") itemId: Long): Response<ApiResponse<WishResponse>>
    
    @GET("api/items/wishes")
    suspend fun getMyWishes(
        @Query("page") page: Int = 0,
        @Query("size") size: Int = 20
    ): Response<ApiResponse<PageResponse<ItemListItem>>>

    // ========== 交换 ==========
    @POST("api/trades")
    suspend fun createTrade(@Body request: CreateTradeRequest): Response<ApiResponse<TradeRequest>>

    @GET("api/trades/{id}")
    suspend fun getTrade(@Path("id") tradeId: Long): Response<ApiResponse<TradeRequest>>

    @PUT("api/trades/{id}/status")
    suspend fun updateTradeStatus(
        @Path("id") tradeId: Long,
        @Body request: UpdateTradeStatusRequest
    ): Response<ApiResponse<TradeRequest>>

    @GET("api/trades/sent")
    suspend fun getSentTrades(
        @Query("page") page: Int = 0,
        @Query("size") size: Int = 20
    ): Response<ApiResponse<PageResponse<TradeRequest>>>

    @GET("api/trades/received")
    suspend fun getReceivedTrades(
        @Query("page") page: Int = 0,
        @Query("size") size: Int = 20
    ): Response<ApiResponse<PageResponse<TradeRequest>>>

    // ========== 聊天 ==========
    @POST("api/chat/send")
    suspend fun sendMessage(@Body request: SendMessageRequest): Response<ApiResponse<Message>>

    @GET("api/chat/conversations")
    suspend fun getConversations(
        @Query("page") page: Int = 0,
        @Query("size") size: Int = 20
    ): Response<ApiResponse<PageResponse<Conversation>>>

    @GET("api/chat/conversations/{id}")
    suspend fun getConversationDetail(
        @Path("id") conversationId: Long,
        @Query("page") page: Int = 0,
        @Query("size") size: Int = 50
    ): Response<ApiResponse<ConversationDetail>>

    // ========== 评价 ==========
    @POST("api/reviews")
    suspend fun createReview(@Body request: CreateReviewRequest): Response<ApiResponse<Review>>

    @GET("api/reviews/user/{userId}")
    suspend fun getUserReviews(
        @Path("userId") userId: Long,
        @Query("page") page: Int = 0,
        @Query("size") size: Int = 20
    ): Response<ApiResponse<PageResponse<Review>>>

    // ========== 管理员 ==========
    @GET("api/admin/config")
    suspend fun getAdminConfig(): Response<ApiResponse<SystemConfigResponse>>

    @POST("api/admin/config/allow-user-chat")
    suspend fun setAllowUserChat(@Body request: AllowRequest): Response<ApiResponse<Unit>>

    @POST("api/admin/config/allow-user-view-items")
    suspend fun setAllowUserViewItems(@Body request: AllowRequest): Response<ApiResponse<Unit>>
}
